src/main/java/com/hillel/items_exchange/service/impl/CategoryServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+package com.hillel.items_exchange.service.impl;
+
+import com.hillel.items_exchange.dao.CategoryRepository;
+import com.hillel.items_exchange.dto.CategoryDto;
+import com.hillel.items_exchange.dto.SubcategoryDto;
+import com.hillel.items_exchange.model.Category;
+import com.hillel.items_exchange.model.Subcategory;
+import com.hillel.items_exchange.service.CategoryService;
+import com.hillel.items_exchange.service.SubcategoryService;
+import java.util.*;
+import lombok.RequiredArgsConstructor;
+import org.modelmapper.ModelMapper;
+import org.springframework.stereotype.Service;
+
+import static com.hillel.items_exchange.mapper.UtilMapper.*;
+
+@Service
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+
+    private final CategoryRepository categoryRepository;
+    private final SubcategoryService subcategoryService;
+    private final ModelMapper modelMapper;
+
+    @Override
+    public List<String> findAllCategoryNames() {
+        return categoryRepository.findAllCategoriesNames();
+    }
+
+    @Override
+    public List<CategoryDto> findAllCategoryDto() {
+        return new ArrayList<>(convertAllTo(categoryRepository.findAll(),
+                CategoryDto.class, ArrayList::new));
+    }
+
+    @Override
+    public Optional<CategoryDto> findCategoryDtoById(long id) {
+        return categoryRepository.findById(id)
+                .map(category -> convertTo(category, CategoryDto.class));
+    }
+
+    @Override
+    public CategoryDto create(CategoryDto categoryDto) {
+        final Category savedCategory = saveCategory(categoryDto);
+        return convertTo(savedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public CategoryDto update(CategoryDto categoryDto) {
+        final Category updatedCategory = saveCategory(categoryDto);
+        return convertTo(updatedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public void removeById(long categoryId) {
+        categoryRepository.deleteById(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoDeletable(long categoryId) {
+        return isCategoryExistsById(categoryId) && isInternalSubcategoriesHaveNotProducts(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoUpdatable(CategoryDto categoryDto) {
+        return isCategoryExistsByIdAndNameOrNotExistsByName(categoryDto.getId(),
+                categoryDto.getName()) && isSubcategoriesExist(categoryDto);
+    }
+
+    @Override
+    public boolean isCategoryDtoValidForCreating(CategoryDto categoryDto) {
+        return isCategoryNameHasNotDuplicate(categoryDto.getName())
+                && categoryDto.getSubcategories().stream().allMatch(this::isSubcategoryIdEqualsZero);
```

Comment:
Because we use 2 cases for Subcategory Ids in this method: if the existing subcategory has been updated, it should exist by id, if a new Subcategory has been added, it has to have ID equals zero.
So we cannot use only one validated annotation for Subcategory id.

Reply:
Exactly in this method isCategoryDtoValidFor**Creating**(CategoryDto categoryDto) you check if ids = 0. 
```suggestion
                ;
```
And add to SubcategoryDto.id `@Zero(groups = New.class, message = "{invalid.new.entity.id}")`